pkg/services/dashboards/service/dashboard_service.go

```diff
@@ -82,11 +94,283 @@ type DashboardServiceImpl struct {
 	k8sclient              client.K8sHandler
 	metrics                *dashboardsMetrics
 	publicDashboardService publicdashboards.ServiceWrapper
+	serverLockService      *serverlock.ServerLockService
+	kvstore                kvstore.KVStore
+
+	dashboardPermissionsReady      chan struct{}
+	k8sDeletedDashboardsCleanupJob *k8sCleanupJob
+}
+
+type k8sCleanupJob struct {
+	stop chan struct{}
+}
+
+func (b *k8sCleanupJob) Stop() {
+	if b.stop != nil {
+		close(b.stop)
+	}
+}
+
+func (dr *DashboardServiceImpl) startK8sDeletedDashboardsCleanupJob() {
+	dr.k8sDeletedDashboardsCleanupJob = &k8sCleanupJob{
+		stop: make(chan struct{}),
+	}
+
+	go func() {
+		ticker := time.NewTicker(k8sDashboardCleanupInterval)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-dr.k8sDeletedDashboardsCleanupJob.stop:
+				return
+			case <-ticker.C:
+				if err := dr.executeCleanupWithLock(context.Background()); err != nil {
+					dr.log.Error("Failed to execute k8s dashboard cleanup", "error", err)
+				}
+			}
+		}
+	}()
+}
+
+func (dr *DashboardServiceImpl) executeCleanupWithLock(ctx context.Context) error {
+	// We're taking a leader-like locking approach here. By locking and executing, but never releasing the lock,
+	// we ensure that other instances of this service can't run in parallel and hence the cleanup will only happen once
+	// per k8sDashboardCleanupInterval by setting the maxInterval and having the time between executions be k8sDashboardCleanupInterval as well.
+	return dr.serverLockService.LockAndExecute(
+		ctx,
+		"k8s_dashboard_cleanup",
+		k8sDashboardCleanupInterval,
+		func(ctx context.Context) {
+			if err := dr.cleanupK8sDashboardResources(ctx, k8sDashboardCleanupBatchSize, k8sDashboardCleanupTimeout); err != nil {
+				dr.log.Error("Failed to cleanup k8s dashboard resources", "error", err)
+			}
+		},
+	)
+}
+
+// cleanupK8sDashboardResources cleans up resources marked for deletion in the k8s API.
+// It processes all organizations, finds dashboards with the trash label, and cleans them up.
+// batchSize specifies how many dashboards to process in a single batch.
+// timeout specifies the timeout duration for the cleanup operation.
+func (dr *DashboardServiceImpl) cleanupK8sDashboardResources(ctx context.Context, batchSize int64, timeout time.Duration) error {
+	ctx, span := tracer.Start(ctx, "dashboards.service.cleanupK8sDashboardResources")
+	defer span.End()
+
+	if !dr.features.IsEnabledGlobally(featuremgmt.FlagKubernetesClientDashboardsFolders) {
```

Comment:
NIT again, but could we elevate this all the way up the call stack and if the feature is disabled not run the cleanup at all?

Reply:
Are feature flags only configurable at startup time or also during runtime? I was assuming that they are configurable at runtime (with some sort of feature flag rollout service) and hence we'd want this check inside the activity as otherwise we'd have no way of starting/stopping already deployed instances if we decide to roll out or roll back the feature flag.

Reply:
Still waiting for a response from Stephanie, but as I looked through the code, it seems that we have some featureflags that require a restart and others do not. I guess in the spirit of making sure that we can be as dynamic as possible (i.e. if excrements hit the fan) in order to turn off a feature without the delay/need of shutting down instances. 
So if you don't mind, I'd just leave this in and going forward always assume that a feature flags state might change at any moment.

Reply:
We're moving toward being able to change feature flags at runtime (historically we have not been able to), for this particular feature flag, we'd need to restart in order to get the clients setup. But as this feature flag will be enabled by default in a few weeks, I'm fine with either approach her.

Reply:
I'll leave it as-is then.